lib/internal/repl.js

```diff
@@ -119,7 +119,7 @@ function setupHistory(repl, historyPath, oldHistoryPath, ready) {
     }
 
     if (data) {
-      repl.history = data.split(/[\n\r]+/, repl.historySize);
+      repl.history = data.split(/[\n\r]+/).slice(repl.historySize * -1);
```

Comment:
I'm not sure how large history files can get, but maybe it'd be more efficient to limit the input to the last `repl.historySize` lines first and then split on that? For example:

```javascript
var reLastLines = new RegExp('(?:.*[\r\n]+){0,' + repl.historySize + '}$');
repl.history = reLastLines.exec(data)[0].match(/[^\r\n]+/g);
```